### PR TITLE
APP-3230: Proxy support

### PR DESCRIPTION
**Pull request by symphony-elias**
### Ticket
[APP-3230](https://perzoinc.atlassian.net/browse/APP-3230)

### Description
Implemented proxy support for webclient and jersey2 clients.

### Dependencies
Depends on a test with a local proxy -> docker-compose.yaml files for running local proxies put in this PR.

### Checklist
- [x] Referenced a ticket in the PR title and in the corresponding section
- [x] Filled properly the description and dependencies, if any
- [x] Unit tests updated or added
- [x] Javadoc added or updated
- [x] Updated the documentation in [docs folder](../docs)

### CODE REVIEW

File: symphony-bdk-http/symphony-bdk-http-jersey2/build.gradle [hunk 1]
```diff
@@ -19,6 +19,7 @@ dependencies {
     implementation 'commons-io:commons-io'
     implementation 'org.apache.commons:commons-lang3'
     implementation 'org.glassfish.jersey.connectors:jersey-apache-connector'
+    implementation 'commons-logging:commons-logging'
```
Comment (symphony-youri) [inline symphony-bdk-http/symphony-bdk-http-jersey2/build.gradle:22]: why do we need it?

File: symphony-bdk-http/symphony-bdk-http-jersey2/src/main/java/com/symphony/bdk/http/jersey2/ApiClientBuilderJersey2.java [hunk 2]
```diff
@@ -153,9 +177,20 @@ private ClientConfig createClientConfig() {
...
+      configureProxy(clientConfig);
+    }
+
     return clientConfig;
   }
 
+  private void configureProxy(ClientConfig clientConfig) {
+    clientConfig.connectorProvider(new ApacheConnectorProvider());
+    clientConfig.property(ClientProperties.PROXY_URI, proxyUrl);
+    clientConfig.property(ClientProperties.PROXY_USERNAME, proxyUser);
```
Comment (symphony-youri) [inline symphony-bdk-http/symphony-bdk-http-jersey2/src/main/java/com/symphony/bdk/http/jersey2/ApiClientBuilderJersey2.java:190]: is that ok to set it to a null value?

**Review by symphony-youri — COMMENTED** [review]
Could you explicit: _Depends on a test with a local proxy put in a dedicated repo_?

File: symphony-bdk-core/src/main/java/com/symphony/bdk/core/config/model/BdkServerConfig.java [hunk 3]
```diff
@@ -12,6 +12,8 @@
   protected static final String DEFAULT_SCHEME = "https";
   protected static final int DEFAULT_HTTPS_PORT = 443;
 
+  protected BdkProxyConfig proxy = null;
```
Comment (symphony-youri) [inline symphony-bdk-core/src/main/java/com/symphony/bdk/core/config/model/BdkServerConfig.java:15]: it is probably already set to null by default

File: symphony-bdk-http/symphony-bdk-http-jersey2/src/main/java/com/symphony/bdk/http/jersey2/ApiClientBuilderJersey2.java (hunk 2 above)
Comment (symphony-elias) [reply symphony-bdk-http/symphony-bdk-http-jersey2/src/main/java/com/symphony/bdk/http/jersey2/ApiClientBuilderJersey2.java:190]: Yes, https://eclipse-ee4j.github.io/jersey.github.io/apidocs/latest/jersey/index.html?org/glassfish/jersey/client/ClientProperties.html says for PROXY_* fields that default value is null.

File: symphony-bdk-http/symphony-bdk-http-jersey2/build.gradle (hunk 1 above)
Comment (symphony-elias) [reply symphony-bdk-http/symphony-bdk-http-jersey2/build.gradle:22]: It is a dependency of jersey-apache-connector, see https://eclipse-ee4j.github.io/jersey.github.io/project-info/2.32/jersey/project/jersey-apache-connector/dependencies.html
So when we are using the `ApacheConnectorProvider`, `AuthMain` fails with `java.lang.NoClassDefFoundError: org/apache/commons/logging/LogFactory`. I can change that to runtimeOnly

File: symphony-bdk-http/symphony-bdk-http-jersey2/src/main/java/com/symphony/bdk/http/jersey2/ApiClientBuilderJersey2.java [hunk 4]
```diff
@@ -144,6 +149,25 @@ public ApiClientBuilder withReadTimeout(int readTimeout) {
     return this;
   }
 
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public ApiClientBuilder withProxy(String proxyHost, int proxyPort) {
+    this.proxyUrl = proxyHost != null ? "http://" + proxyHost + ":" + proxyPort : null;
```
[the remark this answers is not in the document]
Comment (symphony-thibault) [reply symphony-bdk-http/symphony-bdk-http-jersey2/src/main/java/com/symphony/bdk/http/jersey2/ApiClientBuilderJersey2.java:157]: `https` is not a possible value 👍

**Review by symphony-thibault — APPROVED** [review]

**Comment by symphony-thibault** [comment]
I remember that we also talked about providing a `docker-compose.yml` file to easily setup a local proxy. Is that something you can quickly add to the example project or should we create an improvement ticket for later?

File: symphony-bdk-http/symphony-bdk-http-jersey2/build.gradle (hunk 1 above)
Comment (symphony-youri) [reply symphony-bdk-http/symphony-bdk-http-jersey2/build.gradle]: Ok, I could not figure out why it it not come up as a transitive dependency there ... maybe add a comment either inline or the git log explaining that is required by jersey apache connector

**Review by symphony-youri — APPROVED** [review]
Cool 👍

**Review by symphony-youri — APPROVED** [review]

**Pull request merged**